feature_extraction.py

```diff
@@ -140,3 +140,206 @@ def transform(self, X):
+                                  'INCOME_PER_PERSON',
+                                  'PAYMENT_RATE']
+
+    def transform(self, X, y=None):
+        X['DAYS_EMPLOYED'].replace(365243, np.nan, inplace=True)
+        # SIMPLE
+        X['ANNUITY_INCOME_PERCENTAGE'] = X['AMT_ANNUITY'] / X['AMT_INCOME_TOTAL']
+        X['CREDIT_TO_GOODS_RATIO'] = X['AMT_CREDIT'] / X['AMT_GOODS_PRICE']
+        X['DAYS_EMPLOYED_PERCENTAGE'] = X['DAYS_EMPLOYED'] / X['DAYS_BIRTH']
+        X['EXT_SOURCES_MEAN'] = X[['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']].mean(axis=1)
+        X['INCOME_CREDIT_PERCENTAGE'] = X['AMT_INCOME_TOTAL'] / X['AMT_CREDIT']
+        X['INCOME_PER_PERSON'] = X['AMT_INCOME_TOTAL'] / X['CNT_FAM_MEMBERS']
+        X['PAYMENT_RATE'] = X['AMT_ANNUITY'] / X['AMT_CREDIT']
+
+        return {'numerical_features': X[self.application_names]}
+
+
+class Bureau(BaseTransformer):
+    def __init__(self, filepath, id_columns, **kwargs):
+        self.filepath = filepath
+        self.id_columns = id_columns
+        self.bureau_names = ['bureau_active_loans_percentage',
+                             'bureau_average_creditdays_prolonged',
+                             'bureau_average_enddate_future',
+                             'bureau_average_loan_type',
+                             'bureau_credit_enddate_percentage',
+                             'bureau_days_credit_diff',
+                             'bureau_debt_credit_ratio',
+                             'bureau_loan_count',
+                             'bureau_loan_types',
+                             'bureau_overdue_debt_ratio'
+                             ]
+
+    def fit(self, X):
+        bureau = pd.read_csv(self.filepath)
+        bureau['AMT_CREDIT_SUM'].fillna(0, inplace=True)
+        bureau['AMT_CREDIT_SUM_DEBT'].fillna(0, inplace=True)
+        bureau['AMT_CREDIT_SUM_OVERDUE'].fillna(0, inplace=True)
+        bureau['CNT_CREDIT_PROLONG'].fillna(0, inplace=True)
+
+        # NUMBER OF PAST LOANS PER CUSTOMER
+        bureau['bureau_loan_count'] = bureau.groupby(
+            by=['SK_ID_CURR'])['DAYS_CREDIT'].agg('count').reset_index()['DAYS_CREDIT']
+
+        # NUMBER OF TYPES OF PAST LOANS PER CUSTOMER
+        bureau['bureau_loan_types'] = bureau.groupby(
+            by=['SK_ID_CURR'])['CREDIT_TYPE'].agg('nunique').reset_index()['CREDIT_TYPE']
+
+        # AVERAGE NUMBER OF PAST LOANS PER TYPE PER CUSTOMER
+        bureau['bureau_average_loan_type'] = bureau['bureau_loan_count'] / bureau['bureau_loan_types']
+
+        # % OF ACTIVE LOANS FROM BUREAU DATA
+        bureau['bureau_credit_active_binary'] = bureau.apply(lambda x: int(x.CREDIT_ACTIVE != 'Closed'), axis=1)
+        bureau['bureau_active_loans_percentage'] = bureau.groupby(
+            by=['SK_ID_CURR'])['bureau_credit_active_binary'].agg('mean').reset_index()['bureau_credit_active_binary']
+
+        # AVERAGE NUMBER OF DAYS BETWEEN SUCCESSIVE PAST APPLICATIONS FOR EACH CUSTOMER
```

Comment:
@pknut I prefer putting some logic in a method with descriptive name. That makes comments obsolete and makes code easier to read